### ISSUE TITLE
Add UEFI support

### ISSUE THREAD

**Issue by zebradots** [redorescue/redorescue#18]
Is UEFI support required, given that legacy boot options are available in nearly every single machine we've encountered? Is there truly demand for UEFI support, considering only the latest release of Debian 10 added support in its live images for UEFI?

UEFI adds more bloat and complexity to the ISO image creation process and the management of different boot menus, etc., but eventually it will be needed as things continue to move in that direction.

Thoughts and comments welcome.

**Comment by okalm** [comment]
Without going into the "UEFI thing", I don't have enough knowledge for that, knowing that we can and certainly will find more and more usable hardware supported by UEFI than legacy bios is probably one valid point to think about.
Choosing to include UEFI in your project makes more sense in 2020, ten years ago I wouldn't have said this but today things have changed and UEFI is now a "normal thing".
Just like Systemd you must deal with it or you are excluding your project from the most of GNU/Linux's users.
I don't say that nobody will use RedoRescue as it is now, just a lot less than if UEFI would be supported.

But if you decide to go with UEFI, IMO that would be a lot of energy wasted in the same direction for RedoRescue and RescueZilla, I am just saying because I don't know exactly how you two are going to deal with this and if you're doing well together, but that's another story.

Anyway I hope you will sort it out in the good way, and thank you for asking what people think about UEFI support.
Peace & good luck.

**Comment by agnivo007** [comment]
I strongly recommend moving base to debian 10 with UEFI support (Secure boot support is not required as we can easily disable it in the firmware). Notice that most live distros are hybrid UEFI+legacy boot nowadays and must be for a reason and user demand, I don't think including the EFI partition would add 100s of MB anyways... legacy BIOS is going to be obsolete in newer systems very soon; so those users might be cut off from using this tool.

Personally, I have a pendrive loaded with ventoy (www.ventoy.net) where I copy all iso's I need to boot in a folder and it supports UEFI+legacy boot.

**The issue I'm encountering using redorescue on an OLD HP desktop**
(legacy+UEFI support but not so extensive as nowadays):

Due to old BIOS code in the system, which might have limited support for UEFI (latest available bios is installed), if any device has EFI partition, legacy boot option for that partition is greyed out/disabled. When I use the ventoy pendrive, EFI is detected and legacy boot option is not there. Upon booting redorescue iso from ventoy booted into EFI, I get error message that ISO doesn't have a valid  x64 EFI image and hangs up.

One may say why not create a separate pendrive with only redorescue that boots in legacy mode? I, like many other people have troubles juggling between lots of pendrives of various purposes and maintaining them. So, I chose ventoy and added isos for various linux distros, windows iso's and live boot tool iso's like redorescue, like a Swiss-army knife tool.

This issue is not present in newer computers where I can boot ventoy in legacy or EFI mode easily and redorescue loads up and works either way...

**Comment by zebradots** [comment]
There are two available paths forward:

1. Use GRUB for both UEFI and BIOS legacy mode, or
2. Use GRUB for UEFI, and continue to keep using ISOLINUX for BIOS legacy mode

It would be preferable to use GRUB for both, as it requires managing only one menu setup and all the graphic assets that are required for boot screens.

**Comment by agnivo007** [comment]
Are there any specific issues or setbacks if option 1 is used? grub-efi-amd64 should work I guess.

**Comment by zebradots** [comment]
Option one is probably the path of least resistance. We've done a little experimenting to create a live CD that boots on both optical and USB media as both UEFI and BIOS, and there are many moving parts. In some cases it still failed to boot. Those experiments saw the ISO size swell up to add 100M more in size, so hopefully there are ways to mitigate this.

The biggest hurdle at present is determining the exact packages and steps that are required in the `make` script to prepare the image. Once this is complete, we can set about carefully modifying the graphics and menu options to reflect a level of polish consistent with past releases.

**Comment by agnivo007** [comment]
Eager to know if there's any progress...

**Comment by shasheene** [comment]
To mirror [@zebradot's _Redo Rescue_ plug on the Rescuezilla project](https://github.com/rescuezilla/rescuezilla/issues/118#issuecomment-678784705), I should mention the Rescuezilla fork of _Redo Backup and Recovery_ currently supports booting on EFI-only machines (including with Secure Boot enabled). I should also mention that as of writing Rescuezilla is only officially supports whole disk backup/restore so it's not yet well-suited for backup/restore of individual partitions.

Both projects have very different goals and methodologies so are developed almost completely independently.

**Comment by agnivo007** [comment]
Hmm... read the stuff... good if something comes up in UEFI boot code collaboration!

**Comment by zebradots** [comment]
Thanks @shasheene and @agnivo007!

While our testing of Redo has made excellent progress in adding UEFI boot support, the challenges and delays are mostly attributable to the transition to Debian 10. Only with the most recent version (10) did live images support UEFI boot, and current Redo releases are based on Debian 9. So UEFI support will require upgrading the entire image base from Debian 9 to 10, and making sure nothing was broken. It is being worked on, but it will be a little while.

**Comment by agnivo007** [comment]
OK, a wait for the better is worth it... thanks.

**Comment by zebradots** [comment]
Just to provide an update: UEFI support with Secure Boot and the Debian 10 base all function well in testing. The release will be made once all outstanding bugs are corrected for the current major release and we've had time to migrate the changes.

**Comment by zebradots** [comment]
Release 3.0.0 added this feature.

**Comment by MarcelKuesters** [comment]
Hello,

I want to test out Redorescue, but on my iMac 13,1 from 2013 booting is not working. I get only a Windows boot selection.

When I use Rescuezilla on my iMac 13,1 booting is fine. So I made some pictures of the boot structure of booth ISO's.

https://imgur.com/a/UFHGumn
https://imgur.com/U4Hvcyn
https://imgur.com/m6ijd9Z

After some research on the Internet, it looks like that my iMac EFI is case-sensitive.

As we can see in the pictures, everything is in upper-case in Rescuezilla.
Maybe that is a good choice for Redorescue and my iMac will also boot up.

Just some thoughts, I am not an expert in that.

Thanks.

Marcel.
